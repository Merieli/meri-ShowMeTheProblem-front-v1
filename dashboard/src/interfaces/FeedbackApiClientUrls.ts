export interface FeedbackApiClientUrls {
    create: string
    show: string
    getFilters: string
}
