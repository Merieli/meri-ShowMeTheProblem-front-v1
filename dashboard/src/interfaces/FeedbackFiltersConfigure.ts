import { IFeedbackFilters, TFeedback } from './'

export interface FeedbackFiltersConfigure {
    data: IFeedbackFilters
    typeActive: TFeedback
}
