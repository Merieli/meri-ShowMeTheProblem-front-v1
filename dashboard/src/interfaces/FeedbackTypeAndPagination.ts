import { TFeedback } from './'

export interface FeedbackTypeAndPagination {
    type: TFeedback
    limit: number
    offset: number
}
