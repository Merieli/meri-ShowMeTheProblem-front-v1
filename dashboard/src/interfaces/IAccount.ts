export interface IAccount {
    name: {
        value?: string
        errorMessage?: string
    }
    email: {
        value?: string
        errorMessage?: string
    }
    password: {
        value?: string
        errorMessage?: string
    }
}
