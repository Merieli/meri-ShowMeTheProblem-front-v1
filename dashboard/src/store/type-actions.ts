export enum Actions {
    LOGIN_USER = 'LOGIN_USER',
    REGISTER_USER = 'REGISTER_USER',
    GET_USER = 'GET_USER',

    NEW_FEEDBACK = 'NEW_FEEDBACK',
    GET_ALL_FEEDBACKS = 'GET_ALL_FEEDBACKS',
    GET_FEEDBACKS = 'GET_FEEDBACKS',
    GET_INDEX_FEEDBACK = 'GET_INDEX_FEEDBACK',
    CHANGE_ACTIVE_FEEDBACK = 'CHANGE_ACTIVE_FEEDBACK',
}
