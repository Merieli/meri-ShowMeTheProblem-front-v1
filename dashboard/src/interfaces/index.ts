import { IAccount } from './IAccount'

export * from './'
