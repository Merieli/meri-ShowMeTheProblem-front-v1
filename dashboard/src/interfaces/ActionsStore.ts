import { FeedbackTypeAndPagination, IUser, StateStoreShape, TFeedback } from '.';
import { Actions } from '@/store/type-actions';
import { ActionContext, ActionTree } from 'vuex';

type Context = ActionContext<StateStoreShape, StateStoreShape>;

export interface ActionsStore extends ActionTree<StateStoreShape, StateStoreShape> {
    /**
     * @name REGISTER_USER
     * @descripton efetua o registro do usuário e o adiciona a store
     * @param {IUser} user usuário com name, email e password
     */
    [Actions.REGISTER_USER]: ({ commit }: Context, user: IUser) => Promise<void>;

    /**
     * @name LOGIN_USER
     * @description Efetua login do usuário, salva o token na store,
     * redireciona para a página de 'feedbacks' e aciona a action
     * para pegar os dados do usuário logado.
     * @param {IUser} user usuário com name, email e password
     */
    [Actions.LOGIN_USER]: (context: Context, user: IUser) => Promise<void>;

    /**
     * @name GET_USER
     * @descripton Captura os dados do usuário e efetua o login se já tiver um
     * token de usuário salvo, armazenando os dados na store
     */
    [Actions.GET_USER]: (context: Context) => Promise<void>;

    /**
     * @name LOGGOUT_USER
     * @descripton faz o usuário ser deslogado
     */
    [Actions.LOGGOUT_USER]: (context: Context) => Promise<void>;

    /**
     * @name GENERATE_NEW_KEY
     * @descripton gera uma nova chave de api para o usuário e a salva na store
     */
    [Actions.GENERATE_NEW_KEY]: (context: Context) => Promise<void>;

    [Actions.HANDLE_COPY_TEXT]: (context: Context, text: string) => Promise<void>;

    /**
     * @name GET_SUMMARY_FEEDBACK
     * @descripton pega a quantidade de cada tipo de feedback e salva na store
     * configurando para exibir os filtros da listagem
     */
    [Actions.GET_SUMMARY_FEEDBACK]: (context: Context) => Promise<void>;

    /**
     * @name CHANGE_ACTIVE_FEEDBACK
     * @descripton altera o tipo de feedback ativo
     * @param {TFeedback} type tipo do feedback retornado
     */
    [Actions.CHANGE_ACTIVE_FEEDBACK]: (context: Context, type: TFeedback) => Promise<void>;

    /**
     * @name GET_ALL_FEEDBACKS
     * @descripton
     * @param {Object} paramsRequest
     */
    [Actions.GET_ALL_FEEDBACKS]: (
        context: Context,
        feedbackTypeAndPagination: FeedbackTypeAndPagination
    ) => Promise<void>;
}
