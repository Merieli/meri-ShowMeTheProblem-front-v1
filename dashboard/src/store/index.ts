import useNotifier from '../hooks/notifier'
import router from '../router/index'
import { Actions } from './type-actions'
import { Mutations } from './type-mutations'
import callApiClient from '@/http-client/index'
import { TypeOfNotification } from '@/interfaces/INotification'
import { INotification } from '@/interfaces/INotification'
import IUser from '@/interfaces/IUser'
import IUserLogged from '@/interfaces/IUserLogged'
import { createStore } from 'vuex'

export interface EstadoStore {
    users: IUser[]
    notifications: INotification[]
    isLogged: boolean
    userLogged: IUserLogged
}

export const store = createStore<EstadoStore>({
    state: {
        users: [],
        notifications: [],
        isLogged: false,
        userLogged: {
            name: '',
            token: '',
            apiKey: '',
        },
    },
    getters: {},
    mutations: {
        [Mutations.NOTIFY](state, newNotification: INotification) {
            newNotification.id = new Date().getTime()
            state.notifications.push(newNotification)
            setTimeout(() => {
                state.notifications = state.notifications.filter((notif) => notif.id != newNotification.id)
            }, 5000)
        },
        [Mutations.ADD_USER](state, user: IUser) {
            state.users.push(user)
        },
        [Mutations.LOGIN_USER](state, token) {
            state.userLogged.token = token
            state.isLogged = true
        },
        [Mutations.USER_LOGGED](state, data) {
            state.userLogged.name = data.name
            state.userLogged.apiKey = data.apiKey
        },
    },
    actions: {
        /**
         * @name REGISTER_USER
         * @descripton efetua o registro do usuário e o adiciona a store
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.REGISTER_USER]({ commit }, user: IUser) {
            try {
                const response = await callApiClient.user.register(user.name, user.email, user.password)
                commit(Mutations.ADD_USER, response)
            } catch (e) {
                throw new Error(`Não foi possível efetuar cadastro. Confira os dados informados`)
            }
        },

        /**
         * @name LOGIN_USER
         * @description Efetua login do usuário, salva o token na store,
         * redireciona para a página de 'feedbacks' e aciona a action
         * para pegar os dados do usuário logado.
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.LOGIN_USER]({ commit, dispatch, state }, user: IUser) {
            try {
                const response = await callApiClient.user.login(user.email, user.password)
                commit(Mutations.LOGIN_USER, response.token)
                router.push('/feedbacks')
                dispatch(Actions.GET_USER, state.userLogged.token)
            } catch (e) {
                const { notify } = useNotifier()
                notify(TypeOfNotification.FALHA, 'Não foi possível efetuar login', 'Confira os dados informados.')
                throw new Error(`Não foi possível efetuar login. Confira os dados informados`)
            }
        },

        /**
         * @name GET_USER
         * @descripton Captura os dados do usuários e salva na store
         * @param token chave de acesso do usuário
         */
        async [Actions.GET_USER]({ commit }, token) {
            try {
                const response = await callApiClient.user.show(token)
                commit(Mutations.USER_LOGGED, response)
            } catch (e) {
                throw new Error(`Não foi possível capturar os dados do usuário.`)
            }
        },
    },
    modules: {},
})
