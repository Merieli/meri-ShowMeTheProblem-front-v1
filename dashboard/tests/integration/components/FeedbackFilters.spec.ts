import FeedbackFiltres from '@/components/Feedbacks/FeedbackFilters.vue';
import { StateStoreShape } from '@/interfaces';
import { key } from '@/store';
import { Actions } from '@/store/type-actions';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { createStore, Store } from 'vuex';

describe('FeedbackFilters.vue', () => {
    const getters = {
        isLoading: () => mockStore.state.isLoading,
        feedbackFilters: () => mockStore.state.feedbacks.filters,
    };

    const actions = {
        [Actions.GET_SUMMARY_FEEDBACK]: jest.fn(),
        [Actions.CHANGE_ACTIVE_FEEDBACK]: jest.fn(),
    };

    type TStoreState = Pick<StateStoreShape, 'feedbacks' | 'filters' | 'isLoading'>;

    const mockStore: Store<TStoreState> = createStore<TStoreState>({
        state: {
            feedbacks: {
                data: [], //o resultado depende da Action portanto nÃ£o deve ser mockado
                filters: [
                    {
                        label: 'Todos',
                        color: {
                            text: 'text-info',
                            bg: 'bg-info',
                        },
                        amount: 3,
                        active: true,
                        type: 'all',
                    },
                    {
                        label: 'Problemas',
                        color: {
                            text: 'text-danger',
                            bg: 'bg-danger',
                        },
                        amount: 2,
                        active: false,
                        type: 'issue',
                    },
                    {
                        label: 'Ideias',
                        color: {
                            text: 'text-warning',
                            bg: 'bg-warning',
                        },
                        amount: 1,
                        active: false,
                        type: 'idea',
                    },
                    {
                        label: 'Outros',
                        color: {
                            text: 'text-graydark',
                            bg: 'bg-graydark',
                        },
                        amount: 0,
                        active: false,
                        type: 'other',
                    },
                ],
                pagination: {
                    limit: 4,
                    offset: 0,
                    total: 0,
                },
                currentType: 'all',
            },
            filters: {
                all: 3,
                idea: 1,
                issue: 2,
                other: 0,
            },
            isLoading: false,
        },
        getters,
        actions,
    });

    const wrapper: VueWrapper = shallowMount(FeedbackFiltres, {
        global: {
            plugins: [[mockStore, key]],
        },
    });

    describe('IntegraÃ§Ã£o/Componente', () => {
        describe('ðŸ‘€ RenderizaÃ§Ã£o:', () => {
            test('Dado o componente Quando renderizado EntÃ£o deve possuir html equivalente ao Snapshot gravado', () => {
                expect(wrapper.html()).toMatchSnapshot();
            });

            test('Dado os filtros Quando forem renderizados EntÃ£o deve exibir os filtros "Todos, Problemas, Ideias e Outros"', () => {
                const filterLabels = wrapper.findAll('[data-filter="label"]');

                expect(filterLabels[0].text()).toBe('Todos');
                expect(filterLabels[1].text()).toBe('Problemas');
                expect(filterLabels[2].text()).toBe('Ideias');
                expect(filterLabels[3].text()).toBe('Outros');
            });
        });
        // describe('ðŸ§  Comportamento:', () => {})
    });
});
