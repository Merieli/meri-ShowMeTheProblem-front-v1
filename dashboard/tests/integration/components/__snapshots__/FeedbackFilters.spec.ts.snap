// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackFilters.vue IntegraÃ§Ã£o/Componente ğŸ‘€ RenderizaÃ§Ã£o: Dado o componente Quando renderizado EntÃ£o deve possuir html equivalente ao Snapshot gravado 1`] = `""`;
